crates/astria-conductor/src/executor/optimism.rs

```diff
@@ -0,0 +1,111 @@
+//! This module contains functionality for watching for OP-Stack L1 deposit events
+//! and converting them to deposit transactions to be executed on a rollup based on
+//! the OP-Stack.
+//!
+//! It works by querying for deposit events from the `OptimismPortal` contract.
+//! When `populate_rollup_transactions` is called, it queries for all deposit events
+//! since the last call, converts them to deposit transactions, and prepends them to the
+//! other transactions for that rollup block.
+//!
+//! When the deposit transactions are executed on the rollup, it mints the corresponding
+//! funds which were deposited (locked) on the L1 on the L2 to the specified address.
+use std::sync::Arc;
+
+use astria_optimism::{
+    contract::{
+        OptimismPortal,
+        TransactionDepositedFilter,
+    },
+    deposit::convert_deposit_event_to_deposit_tx,
+};
+use color_eyre::eyre::Context;
+use ethers::{
+    prelude::*,
+    types::transaction::eip2718::TypedTransaction,
+};
+use tracing::debug;
+
+use super::{
+    eyre,
+    eyre::Result,
+};
+
+pub(crate) struct Handler {
+    provider: Arc<Provider<Ws>>,
+    optimism_portal_contract: OptimismPortal<Provider<Ws>>,
+    from_block_height: u64,
+}
+
+impl Handler {
+    pub(crate) fn new(
+        ethereum_provider: Arc<Provider<Ws>>,
+        optimism_portal_contract_address: Address,
+        initial_ethereum_l1_block_height: u64,
+    ) -> Self {
+        let optimism_portal_contract = astria_optimism::contract::get_optimism_portal_read_only(
+            ethereum_provider.clone(),
+            optimism_portal_contract_address,
+        );
+
+        Self {
+            provider: ethereum_provider,
+            optimism_portal_contract,
+            from_block_height: initial_ethereum_l1_block_height,
+        }
+    }
+
+    async fn query_events(&mut self) -> Result<Vec<(TransactionDepositedFilter, LogMeta)>> {
+        let to_block = self.provider.get_block_number().await?;
+        let event_filter = self
+            .optimism_portal_contract
+            .event::<TransactionDepositedFilter>()
+            .from_block(self.from_block_height)
+            .to_block(to_block);
+
+        let events = event_filter
+            .query_with_meta()
+            .await
+            .map_err(|e| eyre::eyre!(e))
```

Comment:
I guess that `e` doesn't implement `std::error::Error`?

Reply:
it appears to now 0.0